stdlib/seq.mli

```diff
@@ -0,0 +1,40 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                 Simon Cruanes                                          *)
+(*                                                                        *)
+(*   Copyright 2017 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* Module [Seq]: functional iterators *)
+
+(** @since NEXT_RELEASE *)
+
+type +'a node =
+  | Nil
+  | Cons of 'a * 'a t
+
+and 'a t = unit -> 'a node
```

Comment:
Can you add some documentation to the module?  The general idea ("lazy lists", possibly infinite; should still be deterministic; etc), and a short text for each component.

Some other functions would make sense to add (mapi, map2, exists, for_all, etc).  Do you prefer to keep such discussions for later?  (Having them in this PR take, overall, less energy than adding them later, but I understand you'd prefer this PR to be merged sooner than later.)

Reply:
Make your pick in https://github.com/c-cube/oseq/ :wink: 
I'll add a bit of documentation, but would really like to have this merged soon.

Reply:
I think having these advanced functions later makes sense. The key thing is that we get access to an iterator that can only be made with the ability to access the internals of the stdlib data structures. Specific advanced functions can easily be added, but there's a possibility of some heavy bikeshedding at that point. Now that this mechanism exists, libraries can also take advantage of it.